--- memtest86-3.0/Makefile.orig	2003-12-06 12:44:50.000000000 +0100
+++ memtest86-3.0/Makefile	2003-12-06 13:05:35.338285560 +0100
@@ -21,7 +21,7 @@
 all: memtest.bin memtest
 
 reloc.o: reloc.c
-	$(CC) -c -fPIC -Wall -g -O2 -fno-strict-aliasing reloc.c
+	$(CC) -c $(CCFLAGS) -fPIC -Wall -fno-strict-aliasing reloc.c
 
 test.o: test.c test.h defs.h config.h
 	$(CC) -c $(CCFLAGS) test.c
--- memtest86-3.0/memtest_shared.lds.orig	2002-04-22 08:11:15.000000000 +0200
+++ memtest86-3.0/memtest_shared.lds	2003-12-06 13:35:09.220614456 +0100
@@ -8,6 +8,7 @@
 		_start = .;
 		*(.text)
 		*(.text.*)
+		*(.gnu.linkonce.t.*)
 		*(.plt)
 		_etext = . ;
 	} = 0x9090
